File: KiLCA/tests/test_lorentz_torque.cpp
#include "calc_flre_quants.h"

#include <cmath>
#include <complex>
#include <cstdio>

namespace {
bool close(double actual, double expected, double tolerance = 1.0e-14) {
    return std::abs(actual - expected) <= tolerance;
}

int check(const char* name, double actual, double expected, double tolerance = 1.0e-14) {
    if (close(actual, expected, tolerance)) {
        return 0;
    }

    std::fprintf(stderr, "%s: expected %.17e, got %.17e (difference %.17e)\n", name, expected,
        actual, actual - expected);
    return 1;
}
} // namespace

int main() {
    int failures = 0;

    const std::complex<double> density(1.0, 2.0);
    const std::complex<double> electric_field[3] = {std::complex<double>(3.0, 4.0),
        std::complex<double>(0.0, 0.0), std::complex<double>(0.0, 0.0)};
    const std::complex<double> current_density[3] = {std::complex<double>(1.0, 1.0),
        std::complex<double>(0.0, 0.0), std::complex<double>(0.0, 0.0)};
    const std::complex<double> magnetic_field[3] = {std::complex<double>(0.0, 0.0),
        std::complex<double>(0.0, 0.0), std::complex<double>(2.0, 3.0)};
    double force[3];

    calc_time_averaged_lorentz_force(
        2.0, density, electric_field, current_density, magnetic_field, force);

    failures += check("electrostatic force", force[0], 11.0);
    failures += check("magnetic cross-product sign", force[1], -2.5 / c);
    failures += check("zero force component", force[2], 0.0);

    double torque[3];
    calc_cylindrical_torque_density(5.0, 7.0, force, torque);

    failures += check("poloidal torque moment arm", torque[1], -12.5 / c);
    failures += check("toroidal torque moment arm", torque[2], 0.0);

    double radius[3] = {1.0, 2.0, 4.0};
    double density_profile[3] = {1.0, -2.0, 3.0};
    double integrated[3];

    integrate_over_cylinder(3, radius, density_profile, 10.0, integrated);

    failures += check("integral origin", integrated[0], 0.0);
    failures += check("integral first interval", integrated[1], -15.0);
    failures += check("integral signed total", integrated[2], 65.0);

    if (failures != 0) {
        return 1;
    }

    std::puts("KiLCA Lorentz torque identities passed");
    return 0;
}

File: KiLCA/tests/test_lorentz_torque.f90
! Preserve the Lorentz-force identities introduced on main before the migration.
program test_lorentz_torque
    use, intrinsic :: iso_c_binding, only: c_double
    use kilca_flre_quants_m, only: calc_time_averaged_lorentz_force, &
                                   calc_cylindrical_torque_density, integrate_over_cylinder
    implicit none

    real(c_double), parameter :: cspeed = 29979245800.0d0
    complex(c_double) :: density, electric(3), current(3), magnetic(3)
    real(c_double) :: force(3), torque(3), integrated(3)
    integer :: failures

    failures = 0
    density = (1.0d0, 2.0d0)
    electric = [(3.0d0, 4.0d0), (0.0d0, 0.0d0), (0.0d0, 0.0d0)]
    current = [(1.0d0, 1.0d0), (0.0d0, 0.0d0), (0.0d0, 0.0d0)]
    magnetic = [(0.0d0, 0.0d0), (0.0d0, 0.0d0), (2.0d0, 3.0d0)]
    call calc_time_averaged_lorentz_force(2.0d0, density, electric, current, magnetic, force)
    call check('electrostatic force', force(1), 11.0d0)
    call check('magnetic cross-product sign', force(2) * cspeed, -2.5d0)
    call check('zero force component', force(3), 0.0d0)
    call calc_cylindrical_torque_density(5.0d0, 7.0d0, force, torque)
    call check('radial component', torque(1), 11.0d0)
    call check('poloidal torque moment arm', torque(2) * cspeed, -12.5d0)
    call check('zero toroidal torque', torque(3), 0.0d0)
    call calc_cylindrical_torque_density(5.0d0, 7.0d0, [1.0d0, 2.0d0, -3.0d0], torque)
    call check('toroidal torque moment arm', torque(3), -21.0d0)
    ! A common phase rotation cannot change a time-averaged force.
    call calc_time_averaged_lorentz_force(2.0d0, density * (0.0d0, 1.0d0), &
             electric * (0.0d0, 1.0d0), current * (0.0d0, 1.0d0), magnetic * (0.0d0, 1.0d0), torque)
    call check('phase-invariant electric force', torque(1), force(1))
    call check('phase-invariant magnetic force', torque(2) * cspeed, force(2) * cspeed)
    call integrate_over_cylinder(3, [1.0d0, 2.0d0, 4.0d0], [1.0d0, -2.0d0, 3.0d0], &
                                 10.0d0, integrated)
    call check('integral origin', integrated(1), 0.0d0)
    call check('integral first interval', integrated(2), -15.0d0)
    call check('integral signed total', integrated(3), 65.0d0)
    if (failures /= 0) stop 1
    print *, 'PASS KiLCA Lorentz torque identities'

contains

    subroutine check(label, actual, expected)
        character(len=*), intent(in) :: label
        real(c_double), intent(in) :: actual, expected
        if (abs(actual - expected) <= 1.0d-14) return
        print *, 'FAIL ', label, ': expected', expected, 'got', actual
        failures = failures + 1
    end subroutine check
end program test_lorentz_torque
